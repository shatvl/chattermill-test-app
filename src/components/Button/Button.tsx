import styled from 'styled-components';

export enum ButtonSizes {
  XS = 'xs',
  SM = 'sm',
  MD = 'md',
  LG = 'lg',
}

const Button = styled.button<{
  size?: ButtonSizes | null;
  block?: boolean;
  active?: boolean;
  disabled?: boolean;
}>`
  display: inline-block;
  width: ${({ block }) => (block ? '100%' : 'initial')};
  vertical-align: middle;
  background: ${({ theme, active }) => (active ? theme.colors.primary : 'transparent')};
  border: 1px solid ${({ theme, active }) => (active ? theme.colors.primaryMedium : 'transparent')};
  font-size: ${({ size = ButtonSizes.MD }) => {
    switch (size) {
      case ButtonSizes.XS:
        return '0.875rem';
      case ButtonSizes.SM:
        return '0.875rem';
      case ButtonSizes.MD:
        return '1rem';
      case ButtonSizes.LG:
        return '1.125rem';
      default:
        return '0.875rem';
    }
  }};
  line-height: ${({ size = ButtonSizes.MD }) => {
    switch (size) {
      case ButtonSizes.XS:
        return '1rem';
      case ButtonSizes.SM:
        return '1.25rem';
      case ButtonSizes.MD:
        return '1.5rem';
      case ButtonSizes.LG:
        return '1.75rem';
      default:
        return '1.25rem';
    }
  }};
  padding: ${({ size = ButtonSizes.MD }) => {
    switch (size) {
      case ButtonSizes.XS:
        return '0.25rem 0.375rem';
      case ButtonSizes.SM:
        return '0.5rem 0.525rem';
      case ButtonSizes.MD:
        return '0.625rem 1.25rem';
      case ButtonSizes.LG:
        return '0.75rem 1.25rem';
      default:
        return '0.375rem 1rem';
    }
  }};
  border-radius: 0.25rem;
  color: ${({ theme, disabled }) => (disabled ? `${theme.colors.gray}` : theme.colors.primary)};
  box-sizing: border-box;
  outline: none;
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  transition: 0.15s all ease-in-out;
  text-decoration: none;
  white-space: nowrap;

  &:hover,
  &:active {
    outline: none;
  }

  &:hover {
    background-color: ${({ theme }) => theme.colors.gray200};
    border-color: ${({ theme }) => theme.colors.gray200};
  }

  &:active {
    background-color: ${({ theme }) => theme.colors.primaryMedium};
    border-color: ${({ theme }) => theme.colors.primaryLight};
  }

  &:disabled,
  &:disabled:hover,
  &:disabled:focus,
  &:disabled:active {
    cursor: not-allowed;
    color: ${({ theme }) => theme.colors.gray};
  }
`;

export default Button;
