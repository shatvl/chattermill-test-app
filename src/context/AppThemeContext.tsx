import React, { FC, createContext, useContext } from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

interface AppThemeContextProps {}

export interface AppTheme extends DefaultTheme {
  colors: {
    [key: string]: string;
  };
  zIndex: {
    [key: string]: number;
  };
  fonts: {
    header: string;
    main: string;
  };
  borderWidth: string;
}

// global theme for app, TODO: dark / light mode
export const getTheme = (): AppTheme => ({
  zIndex: {
    xs: 1,
    s: 2,
    m: 3,
    lg: 4,
    xlg: 5,
    xxlg: 6,
  },
  colors: {
    background: '#fff',
    primary: '#ff8489',
    primaryLight: 'rgba(255, 132, 136, .2)',
    primaryMedium: 'rgba(255, 132, 136, .5)',
    red: '#d65942',
    gray: '#c4c9cc',
    grayLight: 'e3e3e3',
    text: '#2e3939',
    textLighter: '#2e3939',
    textSecondary: '#707579',
    textMeta: '#bfbfbf',
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
    header: 'Hind Siliguri',
    main: 'Hind Siliguri',
  },
  borderWidth: '0.0625rem',
});

export const AppThemeContext = createContext<AppThemeContextProps>({});

export const AppThemeProvider: FC = ({ children }) => (
  //we could implement dark / light theme here, or user specific styles base on user session
  <AppThemeContext.Provider value={{}}>
    <ThemeProvider theme={getTheme()}>{children}</ThemeProvider>
  </AppThemeContext.Provider>
);

export const usePortalAppThemeContext = () => useContext(AppThemeContext);
