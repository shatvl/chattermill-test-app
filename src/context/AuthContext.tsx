import React, {
  FC, createContext, useContext, useState, useEffect,
} from 'react';
import jwt_decode from 'jwt-decode';
import { CHATTER_TOKEN_COOKIE, getCoockie } from '../services/cookie';
import { User } from '../models/User.model';

interface AuthContextProps {
  user: User;
  isAuthorized: () => boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: null,
  isAuthorized: () => false,
});

export const AuthProvider: FC = ({ children }) => {
  const [user, setUser] = useState<User>();

  // decode token from params or try to retrieve from cookie
  const getUser = (token?: string): User => {
    const tok = token ?? getCoockie(CHATTER_TOKEN_COOKIE);

    if (!tok) {
      return null;
    }

    try {
      const decodedUser: { exp: number; id: string; orig_iat: number } = jwt_decode(tok);

      return decodedUser || null;
    } catch (error) {
      console.error(error);
      return null;
    }
  };

  const isAuthorized = () =>
    // check via getUser() method, because token could be expired and if users won't decoded navigate to login page
    !!getUser();
  useEffect(() => setUser(getUser()), []);

  return <AuthContext.Provider value={{ user, isAuthorized }}>{children}</AuthContext.Provider>;
};

export const useAuthContext = () => useContext(AuthContext);
