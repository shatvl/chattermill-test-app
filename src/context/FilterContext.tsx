import React, {
  FC, createContext, useContext, useState,
} from 'react';

interface FilterContextProps {
  themeId: number;
  setThemeId: (themeId: number) => void;
  clearFilters: () => void;
}

export const FilterContext = createContext<FilterContextProps>({
  themeId: null,
  setThemeId: () => {},
  clearFilters: () => {},
});

export const FilterProvider: FC = ({ children }) => {
  const [themeId, setThemeId] = useState<number>();

  const handleSelectThemeId = (themeId: number) => setThemeId(themeId);

  const handleClearFilters = () => setThemeId(null);

  return (
    <FilterContext.Provider value={{ themeId, setThemeId: handleSelectThemeId, clearFilters: handleClearFilters }}>
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => useContext(FilterContext);
