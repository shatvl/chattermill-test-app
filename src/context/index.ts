export { AppThemeContext } from './AppThemeContext';
export { usePortalAppThemeContext } from './AppThemeContext';
