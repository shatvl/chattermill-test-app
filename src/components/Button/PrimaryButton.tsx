import styled from 'styled-components';
import Button from './Button';

const PrimaryButton = styled(Button)`
  background: ${({ theme, active }) => (active ? theme.colors.primaryMedium : theme.colors.primary)};
  color: ${({ theme }) => theme.colors.white};
`;

export default PrimaryButton;
