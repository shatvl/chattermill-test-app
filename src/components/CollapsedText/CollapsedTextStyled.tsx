import styled from 'styled-components';
import { Button } from '../Button';

export const CollapsedTextWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  ${Button} {
    margin-left: -0.375rem;
  }
`;

export const WrappedText = styled.div<{ maxHeight: string; lineHeight: string }>`
  max-height: ${({ maxHeight }) => maxHeight};
  line-height: ${({ lineHeight }) => lineHeight};
  width: 100%;
  overflow: hidden;
`;
