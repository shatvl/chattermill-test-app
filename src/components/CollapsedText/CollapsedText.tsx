import React, { FC, useState, useRef, useEffect } from 'react';
import { Button } from '../Button';
import { ButtonSizes } from '../Button/Button';
import { CollapsedTextWrapper, WrappedText } from './CollapsedTextStyled';

const DEFAULT_LINE_HEIGHT = 1.125;
const DEFAULT_VISIBLE_ROWS = 2;
const DEFAULT_FONT_SIZE = 1;

export const calculateLineHeight = (fontSize: number) => (DEFAULT_LINE_HEIGHT / DEFAULT_FONT_SIZE) * fontSize;

interface CollapsedTextProps {
  visibleRows?: number;
  fontSize?: number;
}

/**
 * Component to control `read more` appearence statement (if needed)
 */
const CollapsedText: FC<CollapsedTextProps> = ({ visibleRows = DEFAULT_VISIBLE_ROWS, fontSize = DEFAULT_FONT_SIZE, children }) => {
  const textRef = useRef<HTMLDivElement>();
  const [collapsed, setCollapsed] = useState(false);

  useEffect(() => {
    if (textRef?.current) {
      const textHeight = textRef.current.clientHeight;
      if (textHeight > visibleRows * fontSize * DEFAULT_LINE_HEIGHT * 16) {
        setCollapsed(true);
      }
    }
  }, [fontSize, visibleRows]);

  return (
    <CollapsedTextWrapper>
      <WrappedText
        ref={textRef as any}
        maxHeight={collapsed ? `${calculateLineHeight(fontSize) * visibleRows}rem` : 'initial'}
        lineHeight={`${calculateLineHeight(fontSize)}rem`}
      >
        {children}
      </WrappedText>
      {collapsed && (
        <Button size={ButtonSizes.XS} onClick={() => setCollapsed(!collapsed)}>
          Read more
        </Button>
      )}
    </CollapsedTextWrapper>
  );
};

export default CollapsedText;
