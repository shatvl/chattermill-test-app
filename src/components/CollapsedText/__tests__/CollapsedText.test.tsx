/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import CollapsedText, { calculateLineHeight } from '../CollapsedText';
import { WrappedText, CollapsedTextWrapper } from '../CollapsedTextStyled';
import { wait, waitRenderer } from '../../../testing/wait';
import { Button } from '../../Button';
import { AppThemeProvider } from '../../../context/AppThemeContext';

const originalClientHeight = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'clientHeight');

beforeAll(() => {
  Object.defineProperty(HTMLElement.prototype, 'clientHeight', { configurable: true, value: 100 });
});

afterAll(() => {
  if (originalClientHeight) {
    Object.defineProperty(HTMLElement.prototype, 'clientHeight', originalClientHeight);
  }
});

describe('renders collapsed text', () => {
  it('not collapsed text matches snapshot', async () => {
    const tree = renderer.create(
      <AppThemeProvider>
        <CollapsedText fontSize={1.5}>I&apos;m not collapsed</CollapsedText>
      </AppThemeProvider>,
    );

    await waitRenderer();

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('check not collapsed text', async () => {
    const fontSize = 1.5;
    const visibleRows = 2;
    const component = shallow(
      <CollapsedText visibleRows={visibleRows} fontSize={fontSize}>
        I&apos;m not collapsed
      </CollapsedText>,
    );

    expect(component.find(WrappedText).prop('maxHeight')).toBe('initial');
    expect(component.find(WrappedText).prop('lineHeight')).toBe(`${calculateLineHeight(fontSize)}rem`);
  });

  it('check collapsed text, check click on "read more"', async () => {
    const fontSize = 1.5;
    const visibleRows = 2;
    const component = mount(
      <AppThemeProvider>
        <CollapsedText visibleRows={visibleRows} fontSize={fontSize}>
          Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text.
          Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text.
          Collapsed long text. Collapsed long text. Collapsed long text.
        </CollapsedText>
      </AppThemeProvider>,
    );

    await wait();

    component.update();

    expect(component.find(WrappedText).props().maxHeight).toBe(`${calculateLineHeight(fontSize) * visibleRows}rem`);
    expect(component.find(WrappedText).props().lineHeight).toBe(`${calculateLineHeight(fontSize)}rem`);

    expect(toJson(component.find(CollapsedTextWrapper))).toMatchSnapshot();

    component.find(Button).simulate('click');
    component.update();
    expect(component.find(WrappedText).props().maxHeight).toBe('initial');
  });
});
