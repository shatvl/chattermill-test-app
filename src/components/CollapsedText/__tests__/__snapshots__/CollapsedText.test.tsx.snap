// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders collapsed text check collapsed text, check click on "read more" 1`] = `
.c3 {
  display: inline-block;
  width: initial;
  vertical-align: middle;
  background: transparent;
  border: 1px solid transparent;
  font-size: 0.875rem;
  line-height: 1rem;
  padding: 0.25rem 0.375rem;
  border-radius: 0.25rem;
  color: #ff8489;
  box-sizing: border-box;
  outline: none;
  cursor: pointer;
  -webkit-transition: 0.15s all ease-in-out;
  transition: 0.15s all ease-in-out;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
}

.c3:hover,
.c3:active {
  outline: none;
}

.c3:active {
  background-color: rgba(255,132,136,.5);
  border-color: rgba(255,132,136,.2);
}

.c3:disabled,
.c3:disabled:hover,
.c3:disabled:focus,
.c3:disabled:active {
  cursor: not-allowed;
  color: #c4c9cc;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 .c2 {
  margin-left: -0.375rem;
}

.c1 {
  max-height: 3.375rem;
  line-height: 1.6875rem;
  width: 100%;
  overflow: hidden;
}

<styled.div>
  <div
    className="c0"
  >
    <styled.div
      lineHeight="1.6875rem"
      maxHeight="3.375rem"
    >
      <div
        className="c1"
      >
        Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text. Collapsed long text.
      </div>
    </styled.div>
    <styled.button
      onClick={[Function]}
      size="xs"
    >
      <button
        className="c2 c3"
        onClick={[Function]}
        size="xs"
      >
        Read more
      </button>
    </styled.button>
  </div>
</styled.div>
`;

exports[`renders collapsed text not collapsed text matches snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 .sc-bdnylx {
  margin-left: -0.375rem;
}

.c1 {
  max-height: initial;
  line-height: 1.6875rem;
  width: 100%;
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    I'm not collapsed
  </div>
</div>
`;
